### DISCUSSION TITLE
How to define an interface for a group of value objects?

### DISCUSSION THREAD

**Discussion by glen-84** [SteveDunn/Vogen#328] [Q&A]
I'd like to create a group of value objects representing identifiers, but in order to accept an ID as a parameter, I need some kind of interface to constrain a generic type. What is the best or most correct way to do this (assuming that it's possible)?

**Comment by SteveDunn** [comment]
Hi @glen-84 - sorry, I missed this (I only found it by clicking the wrong button!)

Vogen is designed for wrapping primitives. You could create a composite containing multiple Value Objects. But currently, it's not possible to create a Value Object wrapping multiple fields.

**Reply by glen-84** [reply]
No worries. 🙂

I wasn't looking to create a composite value object, just to apply a common interface to multiple VOs that would allow me to accept any of them as a parameter to a "generic" repository.

However, I think I have some other ideas, so this is no longer important.

Thanks!

**Reply by SteveDunn** [reply]
I see. You could add an interface something like this:

```csharp
    internal interface IHaveAnId<out T>
    {
        public T Value { get; }
    }

    // defaults to int
    [ValueObject]
    internal readonly partial struct CustomerId : IHaveAnId<int>
    {
    }

    // can be internal structs
    [ValueObject]
    internal partial struct AccountId : IHaveAnId<int>
    {
    }
```

I added a new scenario to the `Consumers` project:

https://github.com/SteveDunn/Vogen/blob/main/samples/Vogen.Examples/TypicalScenarios/UsingInterfaces.cs
